### PR TITLE
MLPAB-78 - Only build and deploy new image if /app has changed

### PR DESCRIPTION
**Pull request by andrewpearce-digital**
# Purpose

In order to speed up PR workflow, only build and deploy a new app image when there are changes to /app.

Fixes MLPAB-78

## Approach

- check for changes to /app (and Dockerfile, package.json and yarn.lock) compared to main branch
  - if changes, output changes = true, create a new tag
  - if no changes, output changes = false, pull image tag on production
  - if no changes, Skip unit tests, docker build and ui tests on docker image
  - if no changes, Notify user via Step Summary that jobs are being skipped

image

## Learning

- https://carlosbecker.com/posts/git-changed/
- https://github.blog/2022-05-09-supercharging-github-actions-with-job-summaries/
- https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-job-summary

## Checklist

* [x] I have performed a self-review of my own code
* [x] I have added relevant logging with appropriate levels to my code

### CODE REVIEW

File: .github/workflows/docker_job.yml [hunk 1]
```diff
@@ -36,12 +42,15 @@ jobs:
       - name: Upload Trivy scan results to GitHub Security tab
         id: trivy_upload_sarif
         uses: github/codeql-action/upload-sarif@v2
-        if: always()
+        if: inputs.changes_detected == true
```
Comment (jamesrwarren) [inline .github/workflows/docker_job.yml:45]: you might need to do always() && inputs.changes_detected == true here. If you do this then it won't try and run if changes not detected but will still always run even on failure of previous job. If I remember rightly why it was set to always() in the first place.

Comment (andrewpearce-digital) [reply .github/workflows/docker_job.yml]: that's gone back to always()

File: .github/workflows/tags_job.yml [hunk 2]
```diff
@@ -55,3 +61,42 @@ jobs:
...
+        uses: aws-actions/configure-aws-credentials@v1.7.0
+        with:
+          aws-region: us-east-1
+          role-to-assume: arn:aws:iam::311462405659:role/modernising-lpa-github-actions-ssm-get-parameter
+          role-duration-seconds: 900
+          role-session-name: GithubActionsSSMGetParameter
+      - name: Pull production environment tag
+        id: pull_production_tag
+        run: |
+          value=$( aws ssm get-parameter --name '/modernising-lpa/container-version/preproduction' | jq -r .Parameter.Value )
```
Comment (hawx) [inline .github/workflows/tags_job.yml:79]: pull production tag but gets preproduction?

Comment (andrewpearce-digital) [reply .github/workflows/tags_job.yml:79]: we're not writing a production tag at the moment because it's not on the path to live.

**Review by hawx — APPROVED** [review]

**Review by hawx — APPROVED** [review]

**Pull request merged**